src/quic/connection/sender.rs

```diff
@@ -93,37 +100,67 @@ impl<T: Serialize> Sender<T> {
 	///   stream
 	/// - [`error::Sender::Closed`] if the [`Sender`] is closed
 	#[allow(clippy::panic_in_result_fn, clippy::unwrap_in_result)]
-	pub(super) fn send_any<A: Serialize>(&self, data: &A) -> Result<(), error::Sender> {
+	pub(super) fn send_any<A, AnyFormat>(
+		&self,
+		data: &A,
+		format: &AnyFormat,
+	) -> Result<(), error::Sender>
+	where
+		AnyFormat: Format<'static, A>,
+		AnyFormat::Error: SerializationError,
+	{
 		let mut bytes = BytesMut::new();
 
 		// get size
-		let len = bincode::serialized_size(&data)?;
-		// reserve an appropriate amount of space
-		bytes.reserve(
-			usize::try_from(len)
-				.expect("not a 64-bit system")
-				.checked_add(size_of::<u64>())
-				.expect("data trying to be sent is too big"),
-		);
-		// insert length first, this enables framing
-		bytes.put_u64_le(len);
-
-		let mut bytes = bytes.writer();
-
-		// serialize `data` into `bytes`
-		bincode::serialize_into(&mut bytes, &data)?;
-
-		// send data to task
-		let bytes = bytes.into_inner().freeze();
-
-		// make sure that our length is correct
-		debug_assert_eq!(
-			u64::try_from(bytes.len()).expect("not a 64-bit system"),
-			u64::try_from(size_of::<u64>())
-				.expect("not a 64-bit system")
-				.checked_add(len)
-				.expect("message to long")
-		);
+		let bytes = if let Some(len) = format
+			.serialized_size(data)
+			.map_err(error::Sender::from_serialization)?
+		{
+			// reserve an appropriate amount of space
+			bytes.reserve(
+				len.checked_add(size_of::<u64>())
+					.expect("data trying to be sent is too big"),
+			);
+			// insert length first, this enables framing
+
+			let len = u64::try_from(len).expect("not a 64-bit system");
+			bytes.put_u64_le(len);
+
+			let mut bytes = bytes.writer();
+
+			// serialize `data` into `bytes`
+			format
+				.serialize_into(data, &mut bytes)
+				.map_err(error::Sender::from_serialization)?;
+
+			let bytes = bytes.into_inner().freeze();
+			// make sure that our length is correct
+			debug_assert_eq!(
+				u64::try_from(bytes.len()).expect("not a 64-bit system"),
+				u64::try_from(size_of::<u64>())
+					.expect("not a 64-bit system")
+					.checked_add(len)
+					.expect("message to long")
+			);
+			bytes
+		} else {
+			bytes.put_u64_le(0);
+			let mut bytes = bytes.writer();
+			format
+				.serialize_into(data, &mut bytes)
+				.map_err(error::Sender::from_serialization)?;
+			let mut bytes = bytes.into_inner();
+			let serialized_length = bytes
+				.len()
+				.checked_sub(size_of::<u64>())
+				.expect("negative bytes written");
+			let serialized_length = u64::try_from(serialized_length).expect("not a 64-bit system");
+			bytes
+				.get_mut(0..8)
+				.expect("bytes never allocated")
```

Comment:
I felt really stupid writing this code due to the clippy settings. With this chunk of code, it's crystal clear that this buffer can never be shorter than 8 bytes, yet I'm forced to use the non-panic versions and write `expect` instead of `unwrap()`.

This code went from readable to a mess quickly, and I feel it's in part due to these clippy settings. I'm providing this as feedback for consideration -- not an edict that we should change the settings.